lib/json-schema/attributes/ref.rb

```diff
@@ -21,21 +22,23 @@ def self.validate(current_schema, data, fragments, processor, validator, options
       def self.get_referenced_uri_and_schema(s, current_schema, validator)
         uri,schema = nil,nil
 
-        temp_uri = Addressable::URI.parse(s['$ref'])
+        temp_uri = JSON::Util::URI.parse(s['$ref'])
         if temp_uri.relative?
-          temp_uri = current_schema.uri.clone
+          temp_uri = current_schema.uri.dup
           # Check for absolute path
-          path = s['$ref'].split("#")[0]
+          path, fragment = *s['$ref'].split("#")
           if path.nil? || path == ''
-            temp_uri.path = current_schema.uri.path
+            path = current_schema.uri.path
+            temp_uri.merge!(:path => path, :fragment => fragment)
           elsif path[0,1] == "/"
-            temp_uri.path = Pathname.new(path).cleanpath.to_s
+            path = Pathname.new(path).cleanpath.to_s
+            temp_uri.merge!(:path => path, :fragment => fragment)
           else
             temp_uri = current_schema.uri.join(path)
+            temp_uri.fragment = fragment
           end
-          temp_uri.fragment = s['$ref'].split("#")[1]
         end
-        temp_uri.fragment = "" if temp_uri.fragment.nil?
+        temp_uri = temp_uri.merge(:fragment => "") if temp_uri.fragment.nil? || temp_uri.fragment.empty?
```

Comment:
Let me answer those in turn:

1. If temp_uri.fragment is "", then the URI will have a trailing "#" symbol. If temp_uri.fragment is nil or empty then no "#" is appended. So doing this does change the behaviour
2. We can't use merge! here, because is temp_uri.relative? is false, then temp_uri has come directly from the cache in JSON::Util:URI, and uris stored in the cache are frozen. If you call merge! on a frozen uri it will raise an exception because you are modifying a frozen object. By using merge it will work with both branches of the code